File: sdk-java-blocking/src/test/java/dev/restate/sdk/blocking/SleepTest.java
package dev.restate.sdk.blocking;

import dev.restate.sdk.core.impl.SleepTestSuite;
import dev.restate.sdk.core.impl.testservices.GreeterGrpc;
import dev.restate.sdk.core.impl.testservices.GreetingRequest;
import dev.restate.sdk.core.impl.testservices.GreetingResponse;
import io.grpc.BindableService;
import io.grpc.stub.StreamObserver;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

public class SleepTest extends SleepTestSuite {

  private static class SleepGreeter extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();

      ctx.sleep(Duration.ofMillis(1000));

      responseObserver.onNext(GreetingResponse.newBuilder().setMessage("Hello").build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService sleepGreeter() {
    return new SleepGreeter();
  }

  private static class ManySleeps extends GreeterGrpc.GreeterImplBase
      implements RestateBlockingService {

    @Override
    public void greet(GreetingRequest request, StreamObserver<GreetingResponse> responseObserver) {
      RestateContext ctx = restateContext();
      List<Awaitable<?>> collectedAwaitables = new ArrayList<>();

      for (int i = 0; i < 10; i++) {
        collectedAwaitables.add(ctx.timer(Duration.ofMillis(1000)));
      }

      Awaitable.all(
              collectedAwaitables.get(0),
              collectedAwaitables.get(1),
              collectedAwaitables.subList(2, collectedAwaitables.size()).toArray(Awaitable[]::new))
          .await();

      responseObserver.onNext(GreetingResponse.newBuilder().build());
      responseObserver.onCompleted();
    }
  }

  @Override
  protected BindableService manySleeps() {
    return new ManySleeps();
  }
}

File: sdk-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/SleepTest.kt
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.impl.SleepTestSuite
import dev.restate.sdk.core.impl.testservices.GreeterGrpcKt
import dev.restate.sdk.core.impl.testservices.GreetingRequest
import dev.restate.sdk.core.impl.testservices.GreetingResponse
import dev.restate.sdk.core.impl.testservices.greetingResponse
import io.grpc.BindableService
import kotlin.time.Duration.Companion.milliseconds
import kotlinx.coroutines.Dispatchers

class SleepTest : SleepTestSuite() {
  private class SleepGreeter :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()
      ctx.sleep(1000.milliseconds)
      return greetingResponse { message = "Hello" }
    }
  }

  override fun sleepGreeter(): BindableService {
    return SleepGreeter()
  }

  private class ManySleeps :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateCoroutineService {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = restateContext()
      val awaitables = mutableListOf<Awaitable<Unit>>()
      for (i in 0..9) {
        awaitables.add(ctx.timer(1000.milliseconds))
      }
      awaitables.awaitAll()
      return greetingResponse {}
    }
  }

  override fun manySleeps(): BindableService {
    return ManySleeps()
  }
}
